webengine projection

iCollin opened the pull request:
implement https://github.com/smartdevicelink/generic_hmi/issues/275

this PR is ready for review!

JackLivio commented on src/index.js:
```diff
@@ -78,14 +79,33 @@ class HMIApp extends React.Component {
+                if (mem.usedJSHeapSize / mem.jsHeapSizeLimit > 0.75) {
+                    for (var app of self.props.webEngineApps) {
+                        if (app.runningAppId && app.style.position === 'absolute') {
+                            bcController.onExitApplication('RESOURCE_CONSTRAINT', app.runningAppId);
```
If the app is taking up too much memory, should the HMI also "delete" the iframe? Does that happen indirectly after the OnExitApplication message is sent?

iCollin replied:
This will indirectly remove the iframe after OnAppUnregistered is received by the HMI.

JackLivio commented on src/js/AppStore.js:
```diff
@@ -55,11 +55,25 @@ class AppStore extends React.Component {
+                if (manifest.category === 'WEB_VIEW') {
+                    style = {
+                        position: 'absolute',
+                        width: '960px',
```
The size of the web view needs to take into account the dynamic height and width parameters set in the config and used by the scss files. It might be cleaner to define these styles in an scss file and pass a className within the js code instead.

iCollin replied:
Good point, I applied these changes here: https://github.com/smartdevicelink/generic_hmi/pull/291/commits/ab9eaef5a91b50c9af0b59a089cf96ae4e7909f5

JackLivio commented on src/js/Controllers/UIController.js:
```diff
@@ -295,6 +297,11 @@ class UIController {
     }
     onSystemContext(context, appID) {
+        if (context === 'MAIN' || context === 'ALERT') {
+            store.dispatch(appStoreShowWebView());
```
Should you check that the current route is the appstore/webview before dispatching these actions?

iCollin replied:
You're right, there's an issue here if an app registers as WEB_VIEW and changes templates. That app's iframe would still be shown. I will look into this.

iCollin replied:
In https://github.com/smartdevicelink/generic_hmi/pull/291/commits/a9d1f48c833287bdebb4d097b5e5794e06a7cb57 I added a variable to the state to check whether we were in `/web-view` and used this in index.js alongside this variable that reflected the current context. This worked fine, but I realized that checking if we are in `/web-view` is enough and the context doesn't matter after that. So in https://github.com/smartdevicelink/generic_hmi/pull/291/commits/a1ab7d83384407b8e2808c4727fc9ccfe0bba6e2 I removed the old state variable related to the context.

JackLivio commented on src/js/Templates/WebView/WebView.js:
```diff
@@ -0,0 +1,14 @@
+    render() {
+        return (
+            <div className="webView">
+                <AppHeader backLink="/" menuName="Apps" icon="false"/>
```
I assune WebengineAppContainer was not included in this component because the iframe needs to be running when not in view?

iCollin replied:
Exactly, the iframe has to be outside the route context so it can run when we are in a menu etc.

JackLivio commented on src/js/reducers.js:
```diff
@@ -434,6 +434,9 @@ function ui(state = {}, action) {
                     app.displayLayout = "nonmedia"
                     break
+                case "WEB_VIEW":
+                    app.displayLayout = "webview"
```
I think web-view would be more inline with the naming conventions, except nonmedia is kind of ruining my claim.

iCollin replied:
updated in https://github.com/smartdevicelink/generic_hmi/pull/291/commits/89d551f877e2ce71bde1d3f7092c3ff225a8acf6

If we changed nonmedia we could fix [this special case](https://github.com/smartdevicelink/generic_hmi/pull/291/files#diff-63ab2bf602f05ad65cee8459f425cc94R52) and just use .toLower() for the registerApp call.

JackLivio commented:
FYI style does not persist over ignition cycles for installed projection apps.
![Screen Shot 2020-08-04 at 11 11 21 AM](https://user-images.githubusercontent.com/12038416/89311428-f92f4900-d643-11ea-9aa8-1cd73018a630.png)

iCollin commented:
In https://github.com/smartdevicelink/generic_hmi/pull/291/commits/51ed7e7a33cef0e3aa744a992cc9e7aa15bfe459 I removed webengine app style from the state completely, since iframe visibility will now be controlled by the web-view route modifying `webViewActive` in the state.

I also made it so that all web engine apps will be exited when we are running low on memory instead of just projection apps because I was differentiating them by the style.

Lastly I changed when GetInstalledApps is sent to core so that it won't be sent before the HMI is connected to Core.
